src/Compiler/TypedTree/TypedTreeOps.fs

```diff
@@ -10360,3 +10360,34 @@ let isFSharpExceptionTy g ty =
     | ValueSome tcref -> tcref.IsFSharpException
     | _ -> false
 
+let (|EmptyModuleOrNamespaces|_|) (moduleOrNamespaceContents: ModuleOrNamespaceContents) =
+    match moduleOrNamespaceContents with
+    | TMDefs(defs = defs) ->
+        let mdDefsLength =
+            defs
+            |> List.count (function
+                | ModuleOrNamespaceContents.TMDefRec _
+                | ModuleOrNamespaceContents.TMDefs _ -> true
+                | _ -> false)
+        
+        let emptyModuleOrNamespaces =
+            defs
+            |> List.choose (function
+                | ModuleOrNamespaceContents.TMDefRec _ as defRec
+                | ModuleOrNamespaceContents.TMDefs(defs = [ ModuleOrNamespaceContents.TMDefRec _ as defRec ]) ->
+                    match defRec with
+                    | TMDefRec(bindings = [ ModuleOrNamespaceBinding.Module(mspec, ModuleOrNamespaceContents.TMDefs(defs = defs)) ]) ->
+                        defs
+                        |> List.forall (function
+                            | ModuleOrNamespaceContents.TMDefOpens _
+                            | ModuleOrNamespaceContents.TMDefDo _ -> true
+                            | _ -> false)
+                        |> fun isEmpty -> if isEmpty then Some mspec else None
+                    | _ -> None
+                | _ -> None)
+
+        if mdDefsLength = emptyModuleOrNamespaces.Length then
```

Comment:
There seems to be another edge case which could theoretically happen:

```fsharp
namespace Foo

do ()

namespace Bar

   module Baz =  

       do ()
```

In this case it won't be matched and `namespace Foo` won't be in the signatures.

Reply:
I can live with that, the goal is to avoid purely empty file generation.
This will give a result because of the nested module and that is good enough for my use case.